src/composables/request.js

```diff
@@ -66,86 +63,77 @@ following options:
     return Promise.reject(new Error('file size exceeds limit'));
   }
 
-  if (this.awaitingResponse != null) this.awaitingResponse = true;
+  // eslint-disable-next-line no-param-reassign
+  awaitingResponse.value = true;
 
-  const initialRoute = this.$route;
+  // `router` may be `null` in testing.
+  const initialRoute = router != null ? router.currentRoute.value : null;
   return http.request(withAuth(axiosConfig, requestData.session.token))
     .catch(error => {
-      // this.$route seems to be defined even after the component has been
-      // unmounted.
-      if (this.$route !== initialRoute) throw new Error('route change');
+      if (router != null && router.currentRoute.value !== initialRoute)
+        throw new Error('route change');
 
       if (fulfillProblem != null && error.response != null &&
         isProblem(error.response.data) && fulfillProblem(error.response.data))
         return error.response;
 
-      if (this.awaitingResponse != null) this.awaitingResponse = false;
+      // eslint-disable-next-line no-param-reassign
+      awaitingResponse.value = false;
 
       logAxiosError(logger, error);
-      alert.danger(requestAlertMessage(this.$i18n, error, problemToAlert));
+      alert.danger(requestAlertMessage(i18n, error, problemToAlert));
       throw error;
     })
     .then(response => {
-      if (this.$route !== initialRoute) throw new Error('route change');
-      if (this.awaitingResponse != null) this.awaitingResponse = false;
+      if (router != null && router.currentRoute.value !== initialRoute)
+        throw new Error('route change');
+      // eslint-disable-next-line no-param-reassign
+      awaitingResponse.value = false;
 
       return response;
     });
-}
-
-// @vue/component
-const mixin = {
-  inject: ['container'],
-  watch: {
-    $route() {
-      if (this.awaitingResponse != null) this.awaitingResponse = false;
-    }
-  },
-  methods: {
-    request,
-    post(url, data = undefined, config = undefined) {
-      const full = { ...config, method: 'POST', url };
-      if (data != null) full.data = data;
-      return this.request(full);
-    },
-    put(url, data, config = undefined) {
-      return this.request({ ...config, method: 'PUT', url, data });
-    },
-    patch(url, data, config = undefined) {
-      return this.request({ ...config, method: 'PATCH', url, data });
-    },
-    delete(url, config = undefined) {
-      return this.request({ ...config, method: 'DELETE', url });
-    }
-  }
 };
 
-export default () => mixin;
+export default () => {
+  const container = inject('container');
+  const awaitingResponse = ref(false);
+  const request = _request(container, awaitingResponse);
+
+  const { router } = container;
+  // `router` may be `null` in testing.
+  if (router != null)
+    watch(router.currentRoute, () => { awaitingResponse.value = false; });
+
+  return { request, awaitingResponse: readonly(awaitingResponse) };
```

Comment:
Not super important but I like http helper functions `get`, `post`, 'put`, etc. One thing that we can do is to return an object like:
```
{
  send: _request(container, awaitingResponse),
  post: _request(container, awaitingResponse, 'POST'), // _request should take optional argument for HTTP method
  get: _request(container, awaitingResponse, 'GET'),
  ...
}
```

and then in components we would call `this.request.send` or `this.request.post`, that way we would have context of `get`, `post` and also having `delete` would be possible as well.

Reply:
What do you think about the idea of adding these as properties on the `request()` function? That's what axios does: `axios` is a function with additional properties. That way, you can just call `request()` if you want, but you could also call `this.request.post()`, etc.

Reply:
that would work too

Reply:
This ended up being involved enough that I made a separate PR for it: #765.